### PR TITLE
[FX] Experimental type annotation pass using Python signatures

### PR DESCRIPTION
**Pull request by jamesr66a**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #53832 [FX] Normalize torch. namespace ops
* **#53831 [FX] Experimental type annotation pass using Python signatures**
* #53830 Function to retrieve inspect.Signature instances for PyTorch ops

Use Python function signatures to annotate types for Nodes within an FX graph.
This pulls out Python function signatures for both standard torch.nn Module classes
as well as `torch.nn.functional` functions. Support for `torch.` namespace ops
and tensor methods is forthcoming

Differential Revision: [D26982804](https://our.internmc.facebook.com/intern/diff/D26982804)

### CODE REVIEW

File: torch/fx/experimental/schema_type_annotation.py [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+            kwargs (Dict): Keyword arugments that appear at the callsite for `target`
+
+        Returns:
+
+            Optional[Dict]: Normalized kwargs for `target`, or `None` if this target is not
+                supported
+        """
+        assert not isinstance(target, str)
+        try:
+            sig = inspect.signature(inspect.unwrap(target))
```
Comment (suo) [inline torch/fx/experimental/schema_type_annotation.py:110]: I think by default `inspect.signature` follows wrapped callables, so the `inspect.unwrap` is not necessary.

File: torch/fx/experimental/schema_type_annotation.py [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+            target (Callable): Python callable to normalize arguments for
+            args (Tuple): Arguments that appear at the callsite for `target`
+            kwargs (Dict): Keyword arugments that appear at the callsite for `target`
+
+        Returns:
+
+            Optional[Dict]: Normalized kwargs for `target`, or `None` if this target is not
+                supported
+        """
+        assert not isinstance(target, str)
```
Comment (suo) [inline torch/fx/experimental/schema_type_annotation.py:108]: more appropriately `assert callable(target)` right?

File: torch/fx/experimental/schema_type_annotation.py [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+            maybe_inferred_ts_type = torch._C._jit_try_infer_type(module_itr)
+            if maybe_inferred_ts_type.success():
+                python_type = _torchscript_type_to_python_type(maybe_inferred_ts_type.type())
+                attr_proxy.node.type = python_type if not attr_proxy.node.type else attr_proxy.node.type
+
+        return attr_proxy
+
+    def _extract_python_return_type(self, target : Target) -> Optional[Dict[str, Any]]:
+        """
+        Given a call target, args, and kwargs, return the arguments normalized into
```
Comment (suo) [inline torch/fx/experimental/schema_type_annotation.py:93]: This docstring appears to be out of date. Also the return annotation.

File: torch/fx/experimental/schema_type_annotation.py [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+                # TODO: can we emit the union of these? What are the implications on TorchScript
+                # compilation?
+                if inspect.signature(if_true).return_annotation != inspect.signature(if_false).return_annotation:
+                    return super().call_function(target, args, kwargs)
+                target_for_analysis = if_true
+
+            python_ret_type = self._extract_python_return_type(target_for_analysis)
+
+        return_proxy = super().call_function(target, args, kwargs)
+        return_proxy.node.type = return_proxy.node.type if return_proxy.node.type else python_ret_type
```
Comment (suo) [inline torch/fx/experimental/schema_type_annotation.py:57]: Out of curiosity: if `python_ret_type` was retrieved from a functional, under what conditions would it disagree with the proxy's type? Is it appropriate to always override the Python signature with the proxy type?

File: torch/fx/experimental/schema_type_annotation.py [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+        if self.annotate_get_attrs:
+            module_itr = self.module
+            assert isinstance(target, str)
+            atoms = target.split('.')
+            for i, atom in enumerate(atoms):
+                if not hasattr(module_itr, atom):
+                    raise RuntimeError(f'Node referenced nonextent target {".".join(atoms[:i])}!')
+                module_itr = getattr(module_itr, atom)
+
+            maybe_inferred_ts_type = torch._C._jit_try_infer_type(module_itr)
```
Comment (suo) [inline torch/fx/experimental/schema_type_annotation.py:84]: Why this instead of `type(module_itr)`?

Comment (jamesr66a) [reply torch/fx/experimental/schema_type_annotation.py:84]: Because `type(module_itr)` will return types like `dict` rather than `Dict[K, V]` and that diverges from what the current convention is in FX

File: torch/fx/experimental/schema_type_annotation.py (hunk 4 above)
Comment (jamesr66a) [reply torch/fx/experimental/schema_type_annotation.py:57]: The `Proxy`'s type would have to have been set manually (none of our tools set it for anything other than placeholder and return), so we leave it alone so as to not disturb someone else's custom annotations

**Review by suo — APPROVED** [review]
